source/initial_composition/random_perturbation.cc

```diff
@@ -0,0 +1,118 @@
+/*
+  Copyright (C) 2017 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file doc/COPYING.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/initial_composition/random_perturbation.h>
+
+#include <boost/functional/hash.hpp>
+#include <time.h>
+
+namespace aspect
+{
+  namespace InitialComposition
+  {
+    namespace
+    {
+      template<int dim>
+      std::size_t point_hash(const Point<dim> &position)
+      {
+        std::size_t hash;
+
+        for (unsigned int i = 0; i < dim; ++i)
+          boost::hash_combine(hash,position[i]);
+
+        return hash;
+      }
+    }
+
+    template <int dim>
+    double
+    RandomPerturbation<dim>::
+    initial_composition (const Point<dim> &position,
+                         const unsigned int compositional_index) const
+    {
+      std::size_t seed = point_hash(position);
+      boost::hash_combine(seed,compositional_index);
+      if (use_random_seed)
+        boost::hash_combine(seed,time(NULL));
+
+      random_number_generator.seed(seed);
```

Comment:
This is not thread-safe: if you want to re-use the same random number generator object, you'll have to guard it with a mutex.

Alternatively, you can make the object local to the function.